### PR TITLE
Applying inclined velocity boundary constraint at node

### PR DESCRIPTION
**Pull request by ezrayst**
- [x] add function `apply_inclined_velocity_boundary` in `node` to apply these boundary conditions. I will call this function within `apply_velocity_boundary`
- [x] make test cases for `apply_inclined_velocity_boundary` in `node`

### CODE REVIEW

File: include/node.tcc [hunk 1]
```diff
@@ -272,7 +272,22 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_velocity_constraints() {
...
     const auto phase = static_cast<unsigned>(dir / Tdim);
-    this->velocity_(direction, phase) = constraint.second;
-    this->acceleration_(direction, phase) = 0.;
+
+    if (!generic_velocity_constraints_) {
+      this->velocity_(direction, phase) = constraint.second;
+      this->acceleration_(direction, phase) = 0.;
+    } else {
+      // Transform to local coordinate
+      Eigen::Matrix<double, Tdim, Tnphases> local_velocity =
```
Comment (kks32) [inline include/node.tcc:286]: Pre-compute rotation matrix so we don't have to do inverse twice, but store it here as a const and use it twice.

Comment (kks32) [inline include/node.tcc:286]: `const Eigen::Matrix<double, Tdim, Tdim>  inverse_rotation_matrix_ = rotation_matrix_.inverse();`

File: tests/node_test.cc [hunk 2]
```diff
@@ -776,6 +778,21 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
       for (unsigned i = 0; i < velocity.size(); ++i)
         REQUIRE(node->velocity(Nphase)(i) ==
                 Approx(velocity(i)).epsilon(Tolerance));
+
+      // Apply rotation matrix with Euler angles alpha = 10 deg, beta = 30 deg
```
Comment (kks32) [inline tests/node_test.cc]: Could you create an independent test for general velocity constraints? this test case doesn't show what constraint is being applied? Could you prove that in the specified direction the constraint velocity is zero

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/node.tcc (hunk 1 above)
Comment (ezrayst) [reply include/node.tcc:286]: OK

Comment (ezrayst) [reply include/node.tcc:286]: Is it a private variable? Better that way to avoid inverting the same small matrix over and over?

Comment (kks32) [reply include/node.tcc:286]: Let's not add it to the private variables, because it's more memory. Please change here.

File: include/node.h [hunk 3]
```diff
@@ -207,8 +208,10 @@ class Node : public NodeBase<Tdim> {
   Eigen::Matrix<double, Tdim, Tnphases> acceleration_;
   //! Velocity constraints
   std::map<unsigned, double> velocity_constraints_;
-  //! Rotation matrix for inclined velocity constraints
+  //! Rotation matrix for general velocity constraints
   Eigen::Matrix<double, Tdim, Tdim> rotation_matrix_;
+  //! Inverse rotation matrix for general velocity constraints
+  Eigen::Matrix<double, Tdim, Tdim> inverse_rotation_matrix_;
```
Comment (kks32) [inline include/node.h]: Please don't include it as a private variable.

File: include/node.tcc (hunk 1 above)
Comment (kks32) [reply include/node.tcc:286]: Please do not store it as a private variable.

File: tests/node_test.cc [hunk 4]
```diff
@@ -820,6 +822,88 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
...
+        REQUIRE(node->velocity(Nphase)(i) ==
+                Approx(velocity(i)).epsilon(Tolerance));
+
+      // Check acceleration before constraints
+      acceleration << 5., 5.;
+      for (unsigned i = 0; i < acceleration.size(); ++i)
+        REQUIRE(node->acceleration(Nphase)(i) ==
+                Approx(acceleration(i)).epsilon(Tolerance));
+
+      // Apply velocity constraints
```
Comment (kks32) [inline tests/node_test.cc]: Could you specify this is in the local direction and not globabl coordinate system?

File: tests/node_test.cc [hunk 5]
```diff
@@ -820,6 +822,88 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
...
+      REQUIRE((inverse_rotation_matrix * node->velocity(Nphase))(0) ==
+              Approx(-12.5).epsilon(Tolerance));
+
+      // Check apply constraints
+      acceleration << -0.396139826697847, 0.472101061636807;
+      for (unsigned i = 0; i < Dim; ++i)
+        REQUIRE(node->acceleration(Nphase)(i) ==
+                Approx(acceleration(i)).epsilon(Tolerance));
+
+      // Check that the acceleration is 0 in 0 direction local coordinate
```
Comment (kks32) [inline tests/node_test.cc]:
```suggestion
      // Check that the acceleration is 0 in 0 direction in the local coordinates
```

File: tests/node_test.cc [hunk 6]
```diff
@@ -820,6 +822,88 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
...
+      // Apply constraints
+      node->apply_velocity_constraints();
+
+      // Check apply constraints
+      velocity << -9.583478335521184, -8.025403099849004;
+      for (unsigned i = 0; i < Dim; ++i)
+        REQUIRE(node->velocity(Nphase)(i) ==
+                Approx(velocity(i)).epsilon(Tolerance));
+
+      // Check that the velocity is -12.5 in 0 direction in local coordinate
```
Comment (kks32) [inline tests/node_test.cc]:
```suggestion
      // Check that the velocity is -12.5 in 0 direction in the local coordinate
```

File: tests/node_test.cc [hunk 7]
```diff
@@ -820,6 +822,88 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
...
+      Eigen::Matrix<double, Dim, 1> euler_angles;
+      euler_angles << 10. * M_PI / 180, 30. * M_PI / 180;
+      const auto rotation_matrix = mpm::geometry::rotation_matrix(euler_angles);
+      node->assign_rotation_matrix(rotation_matrix);
+      const auto inverse_rotation_matrix = rotation_matrix.inverse();
+
+      // Apply constraints
+      node->apply_velocity_constraints();
+
+      // Check apply constraints
```
Comment (kks32) [inline tests/node_test.cc]:
```suggestion
      // Check applied velocity constraints in the global coordinates
```

File: tests/node_test.cc [hunk 8]
```diff
@@ -820,6 +822,88 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
...
+      REQUIRE(node->assign_velocity_constraint(0, -12.5) == true);
+
+      // Apply rotation matrix with Euler angles alpha = 10 deg, beta = 30 deg
+      Eigen::Matrix<double, Dim, 1> euler_angles;
+      euler_angles << 10. * M_PI / 180, 30. * M_PI / 180;
+      const auto rotation_matrix = mpm::geometry::rotation_matrix(euler_angles);
+      node->assign_rotation_matrix(rotation_matrix);
+      const auto inverse_rotation_matrix = rotation_matrix.inverse();
+
+      // Apply constraints
```
Comment (kks32) [inline tests/node_test.cc]:
```suggestion
      // Apply inclined velocity constraints
```

File: tests/node_test.cc [hunk 9]
```diff
@@ -820,6 +822,88 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
...
+      velocity << -9.583478335521184, -8.025403099849004;
+      for (unsigned i = 0; i < Dim; ++i)
+        REQUIRE(node->velocity(Nphase)(i) ==
+                Approx(velocity(i)).epsilon(Tolerance));
+
+      // Check that the velocity is -12.5 in 0 direction in local coordinate
+      REQUIRE((inverse_rotation_matrix * node->velocity(Nphase))(0) ==
+              Approx(-12.5).epsilon(Tolerance));
+
+      // Check apply constraints
```
Comment (kks32) [inline tests/node_test.cc]:
```suggestion
      // Check applied constraints on acceleration in the global coordinates
```

File: tests/node_test.cc [hunk 10]
```diff
@@ -1201,5 +1285,94 @@ TEST_CASE("Node is checked for 3D case", "[node][3D]") {
...
+
+      // Check acceleration before constraints
+      acceleration << 5., 5., 5.;
+      for (unsigned i = 0; i < acceleration.size(); ++i)
+        REQUIRE(node->acceleration(Nphase)(i) ==
+                Approx(acceleration(i)).epsilon(Tolerance));
+
+      // Apply velocity constraints
+      REQUIRE(node->assign_velocity_constraint(0, 10.5) == true);
+      REQUIRE(node->assign_velocity_constraint(1, -12.5) == true);
```
Comment (kks32) [inline tests/node_test.cc]: These are not sufficient tests as we are not testing for the third direction. Ideally I'd at least test 2 cases: One with the third dimension unaltered like this case, and one with the constraint in the third dimension.

**Review by kks32 — CHANGES_REQUESTED** [review]

File: tests/node_test.cc (hunk 4 above)
Comment (ezrayst) [reply tests/node_test.cc]: Well the 0 here is the local direction actually. Could you explain more?

File: tests/node_test.cc (hunk 10 above)
Comment (ezrayst) [reply tests/node_test.cc]: I did:
- 1 direction constraint for 2D
- 2 directions constraint for 2D
- 2 directions constraint for 3D
- 3 directions constraint for 3D

File: include/node.tcc [hunk 11]
```diff
@@ -272,7 +272,25 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_velocity_constraints() {
     const auto direction = static_cast<unsigned>(dir % Tdim);
     // Phase: Integer value of division (dir / Tdim)
     const auto phase = static_cast<unsigned>(dir / Tdim);
-    this->velocity_(direction, phase) = constraint.second;
-    this->acceleration_(direction, phase) = 0.;
+
```
Comment (kks32) [inline include/node.tcc:275]: Add a comment to explain:
```
// Velocity constraints are applied on Cartesian boundaries
```

File: include/node.tcc [hunk 12]
```diff
@@ -272,7 +272,25 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_velocity_constraints() {
...
+    if (!generic_velocity_constraints_) {
+      this->velocity_(direction, phase) = constraint.second;
+      this->acceleration_(direction, phase) = 0.;
+    } else {
+      // Compute inverse rotation matrix
```
Comment (kks32) [inline include/node.tcc:282]: `// Velocity constraints on inclined boundaries`

File: tests/node_test.cc [hunk 13]
```diff
@@ -820,6 +822,133 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
...
+        REQUIRE(node->velocity(Nphase)(i) ==
+                Approx(velocity(i)).epsilon(Tolerance));
+
+      // Check acceleration before constraints
+      acceleration << 5., 5.;
+      for (unsigned i = 0; i < acceleration.size(); ++i)
+        REQUIRE(node->acceleration(Nphase)(i) ==
+                Approx(acceleration(i)).epsilon(Tolerance));
+
+      SECTION("Check general velocity constraints in 1 direction") {
```
Comment (kks32) [inline tests/node_test.cc:816]: Could we merge this and the Cartesian constraints, so that they are in independent sections?

**Review by kks32 — CHANGES_REQUESTED** [review]
Please have a call with me before you start making changes.

**Review by kks32 — APPROVED** [review]
